src/transform/src/predicate_pushdown.rs

```diff
@@ -607,6 +619,69 @@ impl PredicatePushdown {
         }
     }
 
+    /// If `s` is of the form
+    /// `(isnull(expr1) && isnull(expr2)) || (expr1 = expr2)`,
+    /// extract `expr1` and `expr2`.
+    fn extract_equal_or_both_null(s: &mut MirScalarExpr) -> Option<(MirScalarExpr, MirScalarExpr)> {
+        // Or, And, and Eq are all commutative functions. For each of these
+        // functions, order expr1 and expr2 so you only need to check
+        // `condition1(expr1) && condition2(expr2)`, and you do
```

Comment:
The ability to convert certain outer joins into inner joins has been lost. This is a full-blown outer join in this branch:

```
materialize=> EXPLAIN SELECT * FROM t1 LEFT JOIN t2 ON ( TRUE ) WHERE t1.f1  = t1.f2 + t2.f2;
                Optimized Plan                
----------------------------------------------
 %0 =                                        +
 | Get materialize.public.t1 (u49)           +
 | Filter !(isnull(#0))                      +
 | ArrangeBy ()                              +
                                             +
 %1 =                                        +
 | Get materialize.public.t2 (u53)           +
                                             +
 %2 =                                        +
 | Join %0 %1 (= #0 (#1 + #3))               +
 | | implementation = Differential %1 %0.()  +
 | | demand = (#0..#3)                       +
                                             +
 %3 =                                        +
 | Get materialize.public.t1 (u49)           +
 | Filter !(isnull(#0)), (null || isnull(#0))+
 | Distinct group=(#0, #1)                   +
 | ArrangeBy (#0, #1)                        +
                                             +
 %4 =                                        +
 | Get materialize.public.t1 (u49)           +
 | ArrangeBy (#1, #0)                        +
                                             +
 %5 =                                        +
 | Join %3 %4 (= #0 #2) (= #1 #3)            +
 | | implementation = DeltaQuery             +
 | |   delta %3 %4.(#1, #0)                  +
 | |   delta %4 %3.(#0, #1)                  +
 | | demand = (#0, #1)                       +
 | Map null, null                            +
 | Project (#0, #1, #4, #5)                  +
                                             +
 %6 =                                        +
 | Union %2 %5                               +
 
(1 row)
```

Whereas with the previous commit it is converted into an inner join:

```
materialize=> EXPLAIN SELECT * FROM t1 LEFT JOIN t2 ON ( TRUE ) WHERE t1.f1  = t1.f2 + t2.f2;
               Optimized Plan               
--------------------------------------------
 %0 =                                      +
 | Get materialize.public.t1 (u4055)       +
 | ArrangeBy ()                            +
                                           +
 %1 =                                      +
 | Get materialize.public.t2 (u4059)       +
                                           +
 %2 = Let l0 =                             +
 | Join %0 %1                              +
 | | implementation = Differential %1 %0.()+
 | | demand = (#0..#3)                     +
                                           +
 %3 =                                      +
 | Get %2 (l0)                             +
 | Filter (#0 = (#1 + #3))                 +
 
(1 row)
```

As t2.f2 can not be ```NULL```  in the final result set as the ```WHERE``` clause will evaluate to false.

Reply:
Technically both plans are cross joins, but I agree the first one is unnecessarily complicated.

Reply:
Is this resolved, or should we look in to this? Generally, the OUTER -> INNER conversion happens with the `nonnull_requirements.rs` transform, which pushes down "you can drop records with NULL columns" information.

Reply:
This has been resolved. Since predicates are properly pushed through lets, the version on the PR branch now looks better than the plan on main (though it's still technically a cross join):

```
materialize=> EXPLAIN SELECT * FROM t1 LEFT JOIN t2 ON ( TRUE ) WHERE t1.f1  = t1.f2 + t2.f2;
               Optimized Plan               
--------------------------------------------
 %0 =                                      +
 | Get materialize.public.t1 (u82)         +
 | ArrangeBy ()                            +
                                           +
 %1 =                                      +
 | Get materialize.public.t2 (u86)         +
                                           +
 %2 =                                      +
 | Join %0 %1 (= #0 (#1 + #3))             +
 | | implementation = Differential %1 %0.()+
 | | demand = (#0..#3)                     +
 | Filter !(isnull(#0))                    +
 
(1 row)
```